core/preconditioner/jacobi.hpp

```diff
@@ -0,0 +1,512 @@
+         * blocks, the array will be transformed into `(x, y, x, y, x)` before
+         * generating the preconditioner. As a consequence, specifying a single
+         * value for this property is exactly equivalent to specifying an array
+         * with a single element set to that value.
+         *
+         * Once an instance of the Jacobi linear operator is generated, the
+         * precisions used for the blocks can be obtained by reading this
+         * property. Whether the parameter was set to a single value or to an
+         * array of values can be queried by reading the
+         * `storage_optimization.is_block_wise` boolean sub-property. If it is
+         * set to `false`, the precision used for all blocks can be obtained
+         * using `storage_optimization.of_all_blocks` or by casting
+         * `storage_optimization` to `precision_reduction`. Independently of the
+         * value of `storage_optimization.is_block_wise`, the
+         * `storage_optimization.block_wise` property will return an array of
+         * precisions used for each block. All values set to
+         * `precision_reduction::autodetect()` will be replaced with the value
+         * representing the precision used for the corresponding block.
+         * If the non-adaptive version of Jacobi is used, the
+         * `storage_optimization.block_wise` Array will be empty.
+         */
+        storage_optimization_type GKO_FACTORY_PARAMETER(
+            storage_optimization, precision_reduction(0, 0));
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Jacobi, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Creates an empty Jacobi preconditioner.
+     *
+     * @param exec  the executor this object is assigned to
+     */
+    explicit Jacobi(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Jacobi>(exec), blocks_(exec)
+    {
+        parameters_.block_pointers.set_executor(exec);
+        parameters_.storage_optimization.block_wise.set_executor(exec);
+    }
+
+    /**
+     * Creates a Jacobi preconditioner from a matrix using a Jacobi::Factory.
+     *
+     * @param factory  the factory to use to create the preconditoner
+     * @param system_matrix  the matrix this preconditioner should be created
+     *                       from
+     */
+    explicit Jacobi(const Factory *factory,
+                    std::shared_ptr<const LinOp> system_matrix)
+        : EnableLinOp<Jacobi>(factory->get_executor(),
+                              transpose(system_matrix->get_size())),
+          parameters_{factory->get_parameters()},
+          storage_scheme_{compute_storage_scheme(parameters_.max_block_size)},
+          num_blocks_{parameters_.block_pointers.get_num_elems() - 1},
+          blocks_(factory->get_executor(),
+                  storage_scheme_.compute_storage_space(
+                      parameters_.block_pointers.get_num_elems() - 1))
```

Comment:
@thoasm here is where the `-1` comes from. If `block_pointers` is empty, than the number of block is actually "unknown" (has to be automatically detected later) and `compute_storage_space` gets  `size_type{} -1` as the total number of blocks.

We can either have the handling inside of `compute_storage_space` or we can move it here and do something like:

```c++
parameters_.block_pointers.get_num_elems() > 0 ?
    storage_scheme_.compute_storage_space(
        parameters_.block_pointers.get_num_elems() - 1) : 0
```

I feel this one is unnecessarily complex - writing a long name twice, using a ternary operator (we could also do it with `if` in a lambda, but it does not really reduce complexity), so I would prefer the other solution, and I'll add a comment for the special case.

Reply:
The current solution is fine, I was just wondering under which circumstances the condition is actually true.